fairscale/optim/oss.py

```diff
@@ -418,58 +418,60 @@ def get_global_rank(group: Any, rank: int) -> int:
-                if global_dst_rank != self.global_rank:
-                    # This request will need to be unrolled
-                    bucket_requests.append((future, dst_rank))
-
-            # Directly broadcast the rest
-            for param in params[i_bucketed:]:
-                direct_requests.append(
-                    dist.broadcast(tensor=param.data, src=global_dst_rank, group=self.group, async_op=True),
+            # All the params are sorted per rank and per increasing size
+            for p in params:
+                # Since all the parameters are already sorted per increasing size, we only need to consider the first ones.
+                if not bucket_sent and offset + p.numel() < buffer_size:
+                    end = offset + p.numel()
+                    buffer[offset:end].copy_(p.data.view(-1))  # type: ignore
+                    bucket_params.append((p, offset, end))
+                    offset = end
+                else:
+                    if offset > 0 and not bucket_sent:
+                        bucket_requests.append(
+                            (
+                                dist.broadcast(tensor=buffer, src=global_src_rank, group=self.group, async_op=True),
+                                src_rank,
+                                bucket_params,
+                            )
+                        )
+
+                        bucket_sent = True
+
+                    direct_requests.append(
+                        dist.broadcast(tensor=p.data, src=global_src_rank, group=self.group, async_op=True)
+                    )
+
+            # Catch a trailing bucket
+            if not bucket_sent:
+                bucket_requests.append(
+                    (
+                        dist.broadcast(tensor=buffer, src=global_src_rank, group=self.group, async_op=True),
+                        src_rank,
+                        bucket_params,
+                    )
                 )
 
         # Unroll the initial packed small parameters
-        for gate, rank in bucket_requests:
-            gate.wait()
-
-            params = per_rank_params[rank]
-            buffer = buffers[rank]
-            i_bucketed = 0  # the number of tensors packed in the buffer
-            offset = 0
-
-            while i_bucketed < len(params) and offset + params[i_bucketed].numel() < buffer_size:
-                end = offset + params[i_bucketed].numel()
-                params[i_bucketed].data.copy_(buffer[offset:end].view_as(params[i_bucketed]))  # type: ignore
-                offset = end
-                i_bucketed += 1
+        for work_handle, src_rank, bucket_params in bucket_requests:
+            work_handle.wait()
+            if src_rank != self.rank:
+                for p, offset, end in bucket_params:
+                    p.data.copy_(buffers[src_rank][offset:end].view_as(p.data))  # type: ignore
```

Comment:
add a stub?

Reply:
Done, you're right I was too lazy, looks better now. Typing and PyTorch is not really a love story